### ISSUE TITLE
CachedReader - returns null

### ISSUE THREAD

**Issue by SpacePossum** [doctrine/annotations#191]
From `2.0` but same on `master`:

https://github.com/doctrine/annotations/blob/2.0/lib/Doctrine/Common/Annotations/CachedReader.php#L74

```php
    public function getClassAnnotations(ReflectionClass $class)
    {
        $cacheKey = $class->getName();

        if (isset($this->loadedAnnotations[$cacheKey])) {
            return $this->loadedAnnotations[$cacheKey];
        }

        if (false === ($annots = $this->fetchFromCache($cacheKey, $class))) {
            $annots = $this->delegate->getClassAnnotations($class);
            $this->saveToCache($cacheKey, $annots);
        }

        return $this->loadedAnnotations[$cacheKey] = $annots;
    }
```

When `$annots` is `false` the method will return `false` while the `Doctrine\Common\Annotations\Reader` interface dictates it must return an `array`. This causes issues @ for example for the `Doctrine\Common\Persistence\Mapping\Driver\AnnotationDriver::isTransient` which tries to iterate `false`.

Is the `false` an exception state, i.e. should an exception being thrown by the reader, or should the reader return an empty `array`, or maybe something else?

**Comment by Ocramius** [comment]
```php
        if (false === ($annots = $this->fetchFromCache($cacheKey, $class))) {
            $annots = $this->delegate->getClassAnnotations($class);
```

`$annots` can only be `false` in that `if()` block, and will be replaced by whatever the `$this->delegate` dictates.

`false` is the normal state for cache misses in `doctrine/cache`.

Unless I'm misunderstanding the issue, this looks invalid due to `$annots` never being `false` when reaching a `return` statement.

**Comment by SpacePossum** [comment]
My report is not as good as intended. 
When inside the `if false` check `$annots` is reassigned `$annots = $this->delegate->getClassAnnotations($class);` of which the result is not type checked. 
So yeah, it seems I jumped to conclusions and the issue is with the `delegate` returning `false`. 
Sadly I've no idea with `delegate` was used and since my report I made changes to the setup making the issue go away (and by that I cannot reproduce).

My guess though, is that `return $this->loadedAnnotations[$key] = include $path;` might need some checking to see if what is read from the cache makes sense.
